feat: ambient egress (#1697)
defenseunicorns/uds-core: docs/reference/configuration/service-mesh/egress.md

---
title: Istio Egress
tableOfContents:
  maxHeadingLevel: 4
---

UDS Core leverages Istio to route dedicated egress out of the service mesh. This document provides an overview and examples of the Istio resources that UDS Core deploys to handle egress.

## Configuring the Egress Workload

### Ambient

> For workloads running in ambient mode, the dedicated egress gateway is a *default* component of UDS Core. It is enabled by default and deploys waypoint workloads to the `istio-egress-waypoint` namespace.

Reviewer: nit: this reads funny to me with the double use of `default`
```suggestion
For workloads running in ambient mode, the dedicated egress gateway is automatically included in UDS Core. It comes pre-enabled and deploys waypoint workloads to the `istio-egress-waypoint` namespace.
```

Additional configurations for the waypoint can be added in the form of helm overrides to the `uds-istio-egress-config` chart in the UDS Bundle, such as:

```yaml
overrides:
  istio-egress-waypoint:
    uds-istio-egress-config:
      values:
        - path: "config.deployment.replicas"
          value: 4
        - path: "config.horizontalPodAutoscaler.minReplicas"
          value: 2
```

> See the values.yaml for additional details and configuration options.

Reviewer 2: nit: link here to the values might be nice?

> 

Reviewer: won't let me make a comment on the line above, we need to add this:
```md
tableOfContents:
  maxHeadingLevel: 4
```

inside of the frontmatter so that the limitation sections are captured in the sidebar navigation.

Author: added

          value: 2
```

> See the [values.yaml](../../../../src/istio/charts/uds-istio-egress-config/values.yaml) for additional details and configuration options.

Reviewer 2: With the docs site this will not link as expected, so it should just be a direct github link on the main branch (will be broken initially, but work once merged).

Wildcards in host names are NOT currently supported.
:::

### Ambient Mode

> The following sample Package CR shows configuring egress to a specific host, "httpbin.org", on port 443. 

Reviewer: nit: remove whitespace

```suggestion
The following sample Package CR shows configuring egress to a specific host, "httpbin.org", on port 443.
```

```yaml
apiVersion: uds.dev/v1alpha1
kind: Package
metadata:
  name: pkg-1
  namespace: egress-gw-1
spec:
  network:
    serviceMesh:
      mode: ambient
    allow:
      - description: "Example Curl"
        direction: Egress
        port: 443

>         remoteHost: httpbin.org

Reviewer 3: Is it possible to use egress for internal services - for example pointing it to a Gitlab deployed inside of the cluster but without using Packages. I'm thinking about large clusters where not everything is controlled by UDS. I'm not sure if this is a valid approach through.

Author: Hm an interesting question... I would assume that if the service was registered in the cluster DNS then that should work. I can try to figure this out if that's something of interest! As for the validity of that approach, I suppose I'd think that was down to what the user was trying to achieve - for instance I'd think it could be valid if they want to track the egressed traffic from their slice of the cluster if it was like a multi-tenant environment

Reviewer 3: Multi-tenant environments are quite interesting! So no pressure, but it would be great to at least know if this approach would work in-cluster or not and document it accordingly.

Author: Ok, got a test set up going and proved that you can access other hosts in-cluster. I'm going to lay out what I did here so you can tell me if you think that's on or off track:
- Deployed modified uds-k3d with 8080 exposed -> added basic uds-slim-dev
- Added a spearate nginx controller / Load balancer component to listen on 8080
- Added an "echo" workload + Ingress for "echo.test" (using nginx-controller)
- Added a curl workload + UDS Package with egress allow to "echo.test", port 80
- Updated cluster coredns with the rewrite of "echo.test" to the nginx-controller service
- Successfully curled the host "echo.test" from my curl workload

If this does accurately mock what you were intending, I can add some additional notes to the egress.md

Reviewer 3: I think this approach is fine. Thanks for checking this. 

As this is a very niche use case, I'm leaving it up to you whether the docs should or should not be added.

        remoteProtocol: TLS
        selector:
          app: curl
        serviceAccount: curl
```

> When a Package CR specifies the `network.allow` field with, at minimum, the `remoteHost` and `port` or `ports` parameters, the UDS Core operator will create the necessary Istio resources to allow traffic to egress from the mesh. For ambient, the `serviceAccount` can be specified for an additional layer of security. The resources that are created include the following:

Reviewer 2: ```suggestion
When a Package CR specifies the `network.allow` field with, at minimum, the `remoteHost` and `port` or `ports` parameters, the UDS Operator will create the necessary Istio resources to allow traffic to egress from the mesh. For ambient, the `serviceAccount` can be specified for an additional layer of security. The resources that are created include the following:
```

small nit, I think we refer to the operator as UDS Operator in other docs

Reviewer 2: Given the below note `If no `serviceAccount` is specified, the `default` service account is used` - I might also rephrase "can be specified for an additional layer of security" to something more explicit about it being needed if your workload is not using the default service account?

* An Istio ServiceEntry, in the package namespace, which is used to define the external service that the workload can access.
* An Istio AuthorizationPolicy, in the package namespace, which is used to enforce that only traffic from workloads using the selected service account can egress. If no `serviceAccount` is specified, the `default` service account is used.
* A shared Istio Waypoint, in the `istio-egress-waypoint` namespace, which is used to route the egress traffic.

#### Limitations

Due to Istio limitations and the selected implementation of shared waypoint egress, if two different egress requests in different packages specify the same host (e.g., pkg1 -> example.com AND pkg2 -> example.com) but request different port or protocols (e.g., pkg1 -> example.com:443/TLS AND pkg2 -> example.com:80/HTTP), the second request will be blocked during reconciliation and the package will fail to reconcile.

### Sidecar Mode

> The following sample Package CR shows configuring egress to a specific host, "httpbin.org", on port 443. 

Reviewer: nit: not part of your changes, but cleanup whitespace

```suggestion
The following sample Package CR shows configuring egress to a specific host, "httpbin.org", on port 443.
```

* Some potential vulnerabilities are introduced using TLS Passthrough - you’ll need to know what’s on the other side of that domain because of [domain fronting](https://en.wikipedia.org/wiki/Domain_fronting) - Essentially, this is only a safe feature for trusted hosts, or hosts you know are not vulnerable to domain fronting

* We are not blocking DNS exfiltration

> ## Debugging Tips

Reviewer 3: Nitpick: Maybe calling troubleshooting would be better?

Author: Good call, that's a much better header. Will update!
